src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Resolve.java

```diff
@@ -3828,6 +3828,51 @@ private List<Type> pruneInterfaces(Type t) {
         return result.toList();
     }
 
+    /**
+     * Determine if an instance field may appear in a constructor prologue.
+     *
+     * <p>
+     * This is only allowed when:
+     *  - The field is being assigned a value (i.e., written but not read)
+     *  - The field is not inherited from a superclass
+     *  - The assignment is not within a lambda, because that would require
+     *    capturing 'this' which is not allowed prior to super().
+     *
+     * <p>
+     * Note, this method doesn't catch all such scenarios, because this method
+     * is invoked for symbol "x" only for "x = 42" but not for "this.x = 42".
+     * To catch those cases, we rely on similar logic in Attr.checkAssignable().
+     */
+    private boolean isAllowedInPrologue(Env<AttrContext> env, VarSymbol v) {
+
+        // Check assumptions
+        Assert.check(env.info.ctorPrologue);
+        Assert.check((v.flags_field & STATIC) == 0);
+
+        // The symbol must appear in the LHS of an assignment statement
+        if (!env.tree.hasTag(ASSIGN))
+            return false;
+
+        // The assignment statement must not be within a lambda
+        if (env.info.isLambda)
+            return false;
+
+        // Get the symbol's qualifier, if any
+        JCExpression lhs = TreeInfo.skipParens(((JCAssign)env.tree).lhs);
+        JCExpression base = lhs.hasTag(SELECT) ? ((JCFieldAccess)lhs).selected : null;
+
+        // The symbol must not be an instance field inherited from a superclass
+        if (v.owner.kind == TYP &&
```

Comment:
The conditions here are eerily similar to those in `checkAssignable`. I wonder if there's a way to unify the codepaths, so that we only have these conditions on one place.

Reply:
Good idea - there can indeed be some consolidation - added in b80e7527ed6.